diff --git a/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/DeleteConflictService.java b/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/DeleteConflictService.java
index e575041cd9..93e364bc93 100644
--- a/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/DeleteConflictService.java
+++ b/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/DeleteConflictService.java
@@ -49,8 +49,8 @@ import java.util.List;
 public class DeleteConflictService {
 	private static final Logger ourLog = LoggerFactory.getLogger(DeleteConflictService.class);
 	public static final int FIRST_QUERY_RESULT_COUNT = 1;
-	public static final int RETRY_QUERY_RESULT_COUNT = 60;
-	public static final int MAX_RETRY_ATTEMPTS = 10;
+	public static final int RETRY_QUERY_RESULT_COUNT = 100;
+	public static final int MAX_RETRY_ATTEMPTS = 100;
 
 	@Autowired
 	DeleteConflictFinderService myDeleteConflictFinderService;
